yaml/src/main/java/com/fasterxml/jackson/dataformat/yaml/YAMLAnchorReplayingParser.java

```diff
@@ -0,0 +1,197 @@
+
+import org.yaml.snakeyaml.LoaderOptions;
+import org.yaml.snakeyaml.events.AliasEvent;
+import org.yaml.snakeyaml.events.Event;
+import org.yaml.snakeyaml.events.MappingEndEvent;
+import org.yaml.snakeyaml.events.MappingStartEvent;
+import org.yaml.snakeyaml.events.NodeEvent;
+import org.yaml.snakeyaml.events.ScalarEvent;
+import org.yaml.snakeyaml.events.CollectionEndEvent;
+import org.yaml.snakeyaml.events.CollectionStartEvent;
+import org.yaml.snakeyaml.nodes.MappingNode;
+
+import com.fasterxml.jackson.core.ObjectCodec;
+import com.fasterxml.jackson.core.io.IOContext;
+
+/**
+ * A parser that remembers the events of anchored parts in yaml and repeats them
+ * to inline these parts when an alias if found instead of only returning an alias.
+ *
+ * Note: this overwrites the getEvent() since the base `super.nextToken()` manages to much state and
+ * it seems to be much simpler to re-emit the events.
+ */
+public class YAMLAnchorReplayingParser extends YAMLParser {
+    private static class AnchorContext {
+        public final String anchor;
+        public final List<Event> events = new ArrayList<>();
+        public int depth = 1;
+
+        public AnchorContext(String anchor) {
+            this.anchor = anchor;
+        }
+    }
+
+    /**
+     *  the maximum number of events that can be replayed
+     */
+    public static final int MAX_EVENTS = 9999;
+
+    /**
+     * the maximum limit of anchors to remember
+     */
+    public static final int MAX_ANCHORS = 9999;
+
+    /**
+     * the maximum limit of merges to follow
+     */
+    public static final int MAX_MERGES = 9999;
+
+    /**
+     * the maximum limit of references to remember
+     */
+    public static final int MAX_REFS = 9999;
+
+    /**
+     * Remembers when a merge has been started in order to skip the corresponding
+     * sequence end which needs to be excluded
+     */
+    private final ArrayDeque<Integer> mergeStack = new ArrayDeque<>();
+
+    /**
+     * Collects nested anchor definitions
+     */
+    private final ArrayDeque<AnchorContext> tokenStack = new ArrayDeque<>();
+
+    /**
+     * Keeps track of the last sequentially found definition of each anchor
+     */
+    private final Map<String, List<Event>> referencedObjects = new HashMap<>();
+
+    /**
+     * Keeps track of events that have been insert when processing alias
+     */
+    private final ArrayDeque<Event> refEvents = new ArrayDeque<>();
+
+    /**
+     * keeps track of the global depth of nested collections
+     */
+    private int globalDepth = 0;
+
+    public YAMLAnchorReplayingParser(IOContext ctxt, int parserFeatures, int formatFeatures, LoaderOptions loaderOptions, ObjectCodec codec, Reader reader) {
+        super(ctxt, parserFeatures, formatFeatures, loaderOptions, codec, reader);
+    }
+
+    private void finishContext(AnchorContext context) {
+        if (referencedObjects.size() + 1 > MAX_REFS) throw new IllegalStateException("too many references in the document");
```

Comment:
These should possibly be `JacksonYAMLParseException` (and methods just declare throwing `IOException`) since caller expects to have to handle those anyway.
This assuming this can happen with legal input (huge doc).

Or even better, `jackson-core` has `StreamConstraintsException`: it is used when size constraints (too deep nesting etc) are violated.